File: Braille-Assistant/app/src/main/java/org/example/brailleassistant/utils/BrailleFilter.java
package org.example.brailleassistant.utils;


import org.opencv.core.Core;

import org.opencv.core.Mat;
import org.opencv.core.MatOfPoint;
import org.opencv.core.Point;
import org.opencv.core.Scalar;
import org.opencv.core.Size;
import org.opencv.imgproc.CLAHE;
import org.opencv.imgproc.Imgproc;

import java.util.ArrayList;

import java.util.List;

import static java.util.Arrays.sort;


public class BrailleFilter {

    public enum FilterType_t {
        SingleSided,
        DoubleSided
    }

    static public class BrailleFilterOutput_t {

        public Mat brailleImage;
        public double[][] brailleDotLocations;

        public BrailleFilterOutput_t(Mat brailleImage, double[][] brailleDotLocations) {
            this.brailleImage = brailleImage;
            this.brailleDotLocations = brailleDotLocations;
        }

    }

    private FilterType_t mFilterType;
    private CLAHE mClaheKernel;
    private Mat mErodeKernel;
    private Mat mDilateKernel;

    private Mat mThresholdImage;
    private Mat mErodedImage;
    private Mat mDilatedImage;
    private Mat mLabels;
    private Mat mStats;
    private Mat mCentroids;

    public BrailleFilter() {

        // Initialize CLAHE Kernel
        mClaheKernel = Imgproc.createCLAHE();
        mClaheKernel.setClipLimit(3);
        Size tileGridSize = new Size(8,8);
        mClaheKernel.setTilesGridSize(tileGridSize);

        // Initialize Erode and Dilate Kernels
        Size erodeKernelSize = new Size(3, 3);
        Size dilateKernelSize = new Size(6, 6);
        mErodeKernel = Imgproc.getStructuringElement(Imgproc.MORPH_ELLIPSE, erodeKernelSize);
        mDilateKernel = Imgproc.getStructuringElement(Imgproc.MORPH_ELLIPSE, dilateKernelSize);

        mThresholdImage = new Mat();
        mErodedImage = new Mat();
        mDilatedImage = new Mat();
        mLabels = new Mat();
        mStats = new Mat();
        mCentroids = new Mat();
    }


    public void setFilterType(FilterType_t filterType) {
        mFilterType = filterType;
    }

    public BrailleFilterOutput_t extractBrailleDots(Mat brailleImage) {

        // perform CLAHE
        Mat lightnessEqualizedImage = this.applyLocalLightnessEqualization(brailleImage);
        // perform GHE
        Mat globallyEqualizedImage = this.applyGlobalHistogramEqualization(lightnessEqualizedImage);
        // threshold image
        BrailleFilterOutput_t filterOutput = this.applyThresholding(globallyEqualizedImage);
        // release memory
        lightnessEqualizedImage.release();
        globallyEqualizedImage.release();
        return filterOutput;

    }

    private Mat applyLocalLightnessEqualization(Mat inputImage) {

        // Convert to LAB colour space and split channels
        Mat labImage = new Mat();
        Imgproc.cvtColor(inputImage, labImage, Imgproc.COLOR_RGB2Lab);
        List<Mat> labChannels = new ArrayList<>();
        Core.split(labImage, labChannels);

        // Apply CLAHE to L-channel and merge channels together
        Mat adjustedLChannel = new Mat();
        mClaheKernel.apply(labChannels.get(0), adjustedLChannel);
        adjustedLChannel.copyTo(labChannels.get(0));
        Core.merge(labChannels, labImage);

        // Convert to BGR colour space
        Mat bgrImage = new Mat();
        Mat grayImage = new Mat();
        Imgproc.cvtColor(labImage, bgrImage, Imgproc.COLOR_Lab2BGR);
        Imgproc.cvtColor(bgrImage, grayImage, Imgproc.COLOR_BGR2GRAY);
        bgrImage.release();

        return grayImage;

    }

    private Mat applyGlobalHistogramEqualization(Mat inputImage) {

        // Apply global histogram equalization and blur
        Mat equalizedImage = new Mat();
        Mat blurredEqualizedImage = new Mat();
        Imgproc.equalizeHist(inputImage, equalizedImage);
        Imgproc.medianBlur(equalizedImage, blurredEqualizedImage, 5);

        return blurredEqualizedImage;
    }

    private BrailleFilterOutput_t applyThresholding(Mat inputImage) {

        // threshold image
        int threshold = calculateThreshold(inputImage);
        Imgproc.threshold(inputImage, mThresholdImage, threshold, 255, Imgproc.THRESH_BINARY_INV);

        // erode noise and then dilate
        Point anchor = new Point(0, 0);
        Imgproc.erode(mThresholdImage, mErodedImage, this.mErodeKernel, anchor,2);
        Imgproc.dilate(mErodedImage, mDilatedImage, this.mDilateKernel, anchor,1);

        // find connected components
        int numLabels = Imgproc.connectedComponentsWithStats(mDilatedImage, mLabels, mStats, mCentroids, 4);
        int componentData[] = new int[5];
        int componentAreas[] = new int[numLabels];
        int totalArea = 0;
        double meanArea;
        mStats.get(0, Imgproc.CC_STAT_AREA, componentAreas);

        // calculate mean area
        for (int i = 1; i < numLabels; i++) { // skip first element (background)
            totalArea += componentAreas[i];
        }
        meanArea = (double) totalArea / (numLabels - 1);

        // colours
        Scalar black = new Scalar(0, 0, 0);
        Scalar white = new Scalar(255, 255, 255);

        // delete smaller elements
        for (int i = 4; i < numLabels; i++) { // skip first element (background)

            mStats.row(i).get(0 ,0, componentData);

            if (componentData[Imgproc.CC_STAT_AREA] < meanArea / 2) {
                Point[] corners = new Point[4];
                corners[0] = new Point(componentData[Imgproc.CC_STAT_LEFT], componentData[Imgproc.CC_STAT_TOP]);
                corners[1] = new Point(componentData[Imgproc.CC_STAT_LEFT] + componentData[Imgproc.CC_STAT_WIDTH], componentData[Imgproc.CC_STAT_TOP]);
                corners[2] = new Point(componentData[Imgproc.CC_STAT_LEFT] + componentData[Imgproc.CC_STAT_WIDTH], componentData[Imgproc.CC_STAT_TOP] + componentData[Imgproc.CC_STAT_HEIGHT]);
                corners[3] = new Point(componentData[Imgproc.CC_STAT_LEFT], componentData[Imgproc.CC_STAT_TOP] + componentData[Imgproc.CC_STAT_HEIGHT]);
                MatOfPoint points = new MatOfPoint(corners);
                Imgproc.fillConvexPoly(mDilatedImage, points, new Scalar(0, 0, 0), 1);

            }

        }

        // find noise free connected components
        numLabels = Imgproc.connectedComponentsWithStats(mDilatedImage, mLabels, mStats, mCentroids, 4);

        double[][] brailleDotLocations = new double[numLabels - 1][2];
        double componentCentre[] = new double[2];

        // replace identified braille dots with circles and extract dot locations
        for (int i = 1; i < numLabels; i++) { // skip first element (background)

            mStats.row(i).get(0, 0, componentData);
            Point[] corners = new Point[4];
            corners[0] = new Point(componentData[Imgproc.CC_STAT_LEFT], componentData[Imgproc.CC_STAT_TOP]);
            corners[1] = new Point(componentData[Imgproc.CC_STAT_LEFT] + componentData[Imgproc.CC_STAT_WIDTH], componentData[Imgproc.CC_STAT_TOP]);
            corners[2] = new Point(componentData[Imgproc.CC_STAT_LEFT] + componentData[Imgproc.CC_STAT_WIDTH], componentData[Imgproc.CC_STAT_TOP] + componentData[Imgproc.CC_STAT_HEIGHT]);
            corners[3] = new Point(componentData[Imgproc.CC_STAT_LEFT], componentData[Imgproc.CC_STAT_TOP] + componentData[Imgproc.CC_STAT_HEIGHT]);

            MatOfPoint points = new MatOfPoint(corners);
            Imgproc.fillConvexPoly(mDilatedImage, points, black, 1);

            mCentroids.row(i).get(0, 0, componentCentre);
            Point centre = new Point(componentCentre[0], componentCentre[1]);
            Imgproc.circle(mDilatedImage, centre, 4, white, 7);

            brailleDotLocations[i-1][0] = componentCentre[0];
            brailleDotLocations[i-1][1] = componentCentre[1];

        }

        // sort dot locations by row order
        sort(brailleDotLocations, new java.util.Comparator<double[]>() {
            public int compare(double[] a, double[] b) {
                return Double.compare(a[1], b[1]);
            }
        });

        return new BrailleFilterOutput_t(mDilatedImage, brailleDotLocations);

    }

    private int calculateThreshold(Mat inputImage) {

        return 20;

    }
}

File: Braille_Algorithm_Development/ImageFilter/BrailleFilter.py
import numpy as np
import cv2


class BrailleFilter(object):

    def __init__(self, filter_type, dot_colour):
        # Print all array values setting
        np.set_printoptions(threshold=np.inf)
        self.mFilterType = filter_type
        self.mDotColour = dot_colour
        self.mStructureElement3 = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3))
        self.mStructureElement6 = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (6, 6))
        self.mClaheKernel = cv2.createCLAHE(clipLimit=3, tileGridSize=(8, 8))

    def __applyLocalLightnessEqualization(self, input_image):
        # Convert image to LAB color model
        image_lab = cv2.cvtColor(input_image, cv2.COLOR_BGR2LAB)
        # Split the image into L, A, and B channels
        l_channel, a_channel, b_channel = cv2.split(image_lab)
        # Create CLAHE Kernel and apply to lightness channel
        adjusted_l_channel = self.mClaheKernel.apply(l_channel)
        # Merge the CLAHE enhanced L channel with the original A and B channel
        merged_channels = cv2.merge((adjusted_l_channel, a_channel, b_channel))
        # Convert image from LAB color model back to RGB color model
        lightness_equalized_image = cv2.cvtColor(merged_channels, cv2.COLOR_LAB2BGR)
        # Convert image to grayscale
        lightness_equalized_image = cv2.cvtColor(lightness_equalized_image, cv2.COLOR_BGR2GRAY)
        return lightness_equalized_image

    def __applyGlobalHistogramEqualization(self, input_image):
        # Calculate histogram of image and fined corresponding CDF
        hist, bins = np.histogram(input_image.flatten(), 256, [0, 256])
        cdf = hist.cumsum()
        cdf_m = np.ma.masked_equal(cdf, 0)
        cdf_m = (cdf_m - cdf_m.min()) * 255 / (cdf_m.max() - cdf_m.min())
        cdf = np.ma.filled(cdf_m, 0).astype('uint8')
        # Apply CDF transform to grayscaled input image
        global_equalized_image = cdf[input_image]
        # Apply median filter to remove high frequency noise
        blurred_global_equalized_image = cv2.medianBlur(global_equalized_image, 7)
        return blurred_global_equalized_image

    def __applyThresholding(self, input_image):
        # White areas indicate the top of each braille dot
        if self.mDotColour == "white":
            ret, white_image = cv2.threshold(input_image, 246, 255, cv2.THRESH_BINARY)
        elif self.mDotColour == "black":
            ret, white_image = cv2.threshold(input_image, 15, 255, cv2.THRESH_BINARY)
        else:
            exit()
        eroded_white_image = cv2.erode(white_image, self.mStructureElement3, iterations=2)
        dilated_white_image = cv2.dilate(eroded_white_image, self.mStructureElement6, iterations=1)
        if self.mDotColour == "black":
            dilated_white_image = cv2.bitwise_not(dilated_white_image)

        # Find connected components and extract the mean area
        cc_output = cv2.connectedComponentsWithStats(dilated_white_image, 4, cv2.CV_8U)
        stats = cc_output[2]
        stats = np.delete(stats, 0, 0)  # delete background component
        mean_area = np.mean(stats[:, cv2.CC_STAT_AREA])

        # Remove components that are < mean_area/1.5
        for stat in stats:
            if stat[cv2.CC_STAT_AREA] < mean_area / 2:
                x1 = stat[cv2.CC_STAT_LEFT]
                x2 = x1 + stat[cv2.CC_STAT_WIDTH]
                y1 = stat[cv2.CC_STAT_TOP]
                y2 = y1 + stat[cv2.CC_STAT_HEIGHT]
                cv2.rectangle(dilated_white_image, (x1, y1), (x2, y2), 0, cv2.FILLED)

        cc_output = cv2.connectedComponentsWithStats(dilated_white_image, 4, cv2.CV_8U)
        stats = cc_output[2]
        centroids = cc_output[3]
        braille_dot_stats = np.delete(stats, 0, 0)  # delete background component
        braille_dot_centres = np.delete(centroids, 0, 0)  # delete background component
        braille_dot_locations_list = []

        for stat, dot in zip(braille_dot_stats, braille_dot_centres):
            x1 = stat[cv2.CC_STAT_LEFT]
            x2 = x1 + stat[cv2.CC_STAT_WIDTH]
            y1 = stat[cv2.CC_STAT_TOP]
            y2 = y1 + stat[cv2.CC_STAT_HEIGHT]
            cv2.rectangle(dilated_white_image, (x1, y1), (x2, y2), 0, cv2.FILLED)
            cv2.circle(dilated_white_image, (int(dot[0]), int(dot[1])), 3, 255, cv2.FILLED)
            braille_dot_locations_list.append([int(dot[0]), int(dot[1])])

        # sort dot locations based on row
        braille_dot_locations = np.asarray(braille_dot_locations_list)
        if len(braille_dot_locations) > 1:
            braille_dot_locations = braille_dot_locations[braille_dot_locations[:, 1].argsort(kind='mergesort')]

        return dilated_white_image, braille_dot_locations

    def extractBrailleDots(self, input_image):
        if self.mFilterType == "Single-Sided":
            lightness_equalized_image = self.__applyLocalLightnessEqualization(input_image)
            global_equalized_image = self.__applyGlobalHistogramEqualization(lightness_equalized_image)
            output_image, braille_dot_locations = self.__applyThresholding(global_equalized_image)
            return output_image, braille_dot_locations
